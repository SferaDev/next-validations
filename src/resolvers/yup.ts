import * as Yup from 'yup';

export function yupResolver<T extends Yup.AnyObjectSchema>(schema: T) {
  return {
    validate: (data: unknown) => schema.validateSync(data),
  };
}
