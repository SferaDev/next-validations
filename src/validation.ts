/* eslint-disable @typescript-eslint/naming-convention */
import type { LiteralUnion } from "type-fest";

import { fastestValidatorResolver } from "./resolvers/fastest-validator";
import { joiResolver } from "./resolvers/joi";
import { yupResolver } from "./resolvers/yup";
import { zodResolver } from "./resolvers/zod";

export type SCHEMA_TYPE = LiteralUnion<
  | "Yup"
  | "FastestValidator"
  | "Joi"
  | "Zod",
  string
>;

export function createResolver(type: SCHEMA_TYPE, schema: any) {
  switch (type) {
    case "Yup":
      return yupResolver(schema);
    case "FastestValidator":
      return fastestValidatorResolver(schema);
    case "Joi":
      return joiResolver(schema);
    case "Zod":
      return zodResolver(schema);

    default:
      throw new Error(`Does not support ${type} validation yet!`);
  }
}
