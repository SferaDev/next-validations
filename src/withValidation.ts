import { NextApiRequest, NextApiResponse } from "next";

import { createResolver } from "./validation";
import type { SCHEMA_TYPE } from "./validation";

type NextHandler = (err?: Error) => void;

 type ValidationHoF = {
  type: SCHEMA_TYPE;
  mode?: "body" | "query" | "headers";
  schema: unknown;
};

export function withValidation({
  type,
  schema,
  mode = "query",
}: ValidationHoF) {
  return (
    handler?: (req: NextApiRequest, res: NextApiResponse<any>) => any,
  ) => {
    return async (
      req: NextApiRequest,
      res: NextApiResponse,
      next?: NextHandler,
    ) => {
      try {
        const resolver = createResolver(type, schema);

        resolver.validate(req[mode]);

        if (!!next) {
          return next();
        }

        if (handler) return handler(req, res);

        res.status(404).end();
      } catch (error) {
        res.status(400).send(error);
      }
    };
  };
}

export function withValidations(validations: ValidationHoF[]) {
  return (
    handler?: (req: NextApiRequest, res: NextApiResponse<any>) => any,
  ) => {
    return async (
      req: NextApiRequest,
      res: NextApiResponse,
      next?: NextHandler,
    ) => {
      try {
        validations.forEach((validation) => {
          const resolver = createResolver(validation.type, validation.schema);
          resolver.validate(req[validation.mode || "query"]);
        });

        if (!!next) {
          return next();
        }

        if (handler) return handler(req, res);

        res.status(404).end();
      } catch (error) {
        res.status(400).send(error);
      }
    };
  };
}
