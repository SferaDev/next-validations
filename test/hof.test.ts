import * as yup from "yup";
import { withValidation, withValidations } from "../src";

const schema = yup.object().shape({
  name: yup.string().required(),
});

describe("withValidation", () => {
  it("should return function", () => {
    expect(withValidation({
      schema,
      type: "Yup",
      mode: "query",
    })).toMatchSnapshot();
  });
});

it("should handle App Router validation errors", () => {
  const validation = {
    schema,
    type: "Yup",
    mode: "query",
    apiType: "appRoute",
  };
  const handler = jest.fn();
  const res = {
    status: jest.fn().mockReturnThis(),
    end: jest.fn(),
  };
  withValidation(validation)(null, res, handler);
  expect(res.status).toHaveBeenCalledWith(404);
  expect(res.end).toHaveBeenCalled();
});

it("should handle traditional API route validation errors", () => {
  const validation = {
    schema,
    type: "Yup",
    mode: "query",
    apiType: "pageRoute",
  };
  const handler = jest.fn();
  const res = {
    status: jest.fn().mockReturnThis(),
    send: jest.fn(),
  };
  const error = new Error("Validation error");
  withValidation(validation)(error, null, res, handler);
  expect(res.status).toHaveBeenCalledWith(400);
  expect(res.send).toHaveBeenCalledWith(error.message);
});

describe("withValidations", () => {
  it("should return function", () => {
    expect(withValidations([{
      schema,
      type: "Yup",
      mode: "body",
    }])).toMatchSnapshot();
  });
});

it("should handle App Router validation errors in withValidations", () => {
  const validations = [{
    schema,
    type: "Yup",
    mode: "query",
    apiType: "appRoute",
  }];
  const handler = jest.fn();
  const res = {
    status: jest.fn().mockReturnThis(),
    end: jest.fn(),
  };
  withValidations(validations)(null, res, handler);
  expect(res.status).toHaveBeenCalledWith(404);
  expect(res.end).toHaveBeenCalled();
});

it("should handle traditional API route validation errors in withValidations", () => {
  const validations = [{
    schema,
    type: "Yup",
    mode: "query",
    apiType: "pageRoute",
  }];
  const handler = jest.fn();
  const res = {
    status: jest.fn().mockReturnThis(),
    send: jest.fn(),
  };
  const error = new Error("Validation error");
  withValidations(validations)(error, null, res, handler);
  expect(res.status).toHaveBeenCalledWith(400);
  expect(res.send).toHaveBeenCalledWith(error.message);
});