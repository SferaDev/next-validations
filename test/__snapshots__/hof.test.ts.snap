// Vitest Snapshot v1

exports[`withValidation > should return function 1`] = `[Function]`;

exports[`withValidations > should return function 1`] = `[Function]`;
